### PR TITLE
Add nitpicky mode to sphinx.

### PR DESCRIPTION
**Pull request by hategan**
This addresses part of #109 in that various intersphinx links are now checked and the docs don't build if the links are broken.

However, this is not sufficient, since it does not check other things such as the name of the documented parameters against a method signature. Nonetheless, a big step forward, as can be seen from the changes.

### CODE REVIEW

**Review by andre-merzky — CHANGES_REQUESTED** [review]
The changes seem all to make sense to me, but it's a fairly large diff, hope I did not miss anything.  However, the document  build issues quite a number of warnings for me.  Steps taken:
```sh
$ pip install -r requirements.txt
$ pip install -r requirements-docs.txt
$ make docs
rm -rf docs/.generated
sphinx-build --keep-going -n -W -b html docs docs/.build/
Running Sphinx v4.2.0
/mnt/home/merzky/projects/exaworks/psi-j-python/src/
Could not import /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij-descriptors/aprun_descriptor.py: A launcher by the name "aprun" with version 0.0.1 is already registered. Existing path: /mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/launchers/aprun.py; current path: /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij/launchers/aprun.py
Could not import /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij-descriptors/cobalt_descriptor.py: An executor by the name "cobalt" with version 0.0.1 is already registered. Existing path: /mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/batch/cobalt.py; current path: /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij/executors/batch/cobalt.py
Could not import /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij-descriptors/core_descriptors.py: An executor by the name "local" with version 0.0.1 is already registered. Existing path: /mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/local.py; current path: /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij/executors/local.py
Could not import /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij-descriptors/jsrun_descriptor.py: A launcher by the name "jrun" with version 0.0.1 is already registered. Existing path: /mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/launchers/jsrun.py; current path: /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij/launchers/jsrun.py
Could not import /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij-descriptors/lsf_descriptor.py: An executor by the name "lsf" with version 0.0.1 is already registered. Existing path: /mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/batch/lsf.py; current path: /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij/executors/batch/lsf.py
Could not import /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij-descriptors/pbspro_descriptor.py: An executor by the name "pbspro" with version 0.0.1 is already registered. Existing path: /mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/batch/pbspro.py; current path: /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij/executors/batch/pbspro.py
Could not import /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij-descriptors/rp_descriptor.py: An executor by the name "rp" with version 0.0.1 is already registered. Existing path: /mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/rp.py; current path: /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij/executors/rp.py
Could not import /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij-descriptors/saga_descriptor.py: An executor by the name "saga" with version 0.0.1 is already registered. Existing path: /mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/saga.py; current path: /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij/executors/saga.py
Could not import /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij-descriptors/slurm_descriptor.py: An executor by the name "slurm" with version 0.0.1 is already registered. Existing path: /mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/batch/slurm.py; current path: /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij/executors/batch/slurm.py
Could not import /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij-descriptors/srun_descriptor.py: A launcher by the name "srun" with version 0.0.1 is already registered. Existing path: /mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/launchers/srun.py; current path: /mnt/home/merzky/projects/exaworks/psi-j-python/ve3/lib/python3.8/site-packages/psij/launchers/srun.py
loading pickled environment... done
OUTPUT PATH = /mnt/home/merzky/projects/exaworks/psi-j-python/docs/.generated
Creating file /mnt/home/merzky/projects/exaworks/psi-j-python/docs/.generated/psij.rst.
Creating file /mnt/home/merzky/projects/exaworks/psi-j-python/docs/.generated/psij.executors.rst.
Creating file /mnt/home/merzky/projects/exaworks/psi-j-python/docs/.generated/psij.executors.batch.rst.
Creating file /mnt/home/merzky/projects/exaworks/psi-j-python/docs/.generated/psij.launchers.rst.
Creating file /mnt/home/merzky/projects/exaworks/psi-j-python/docs/.generated/psij.launchers.scripts.rst.
Creating file /mnt/home/merzky/projects/exaworks/psi-j-python/docs/.generated/modules.rst.
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 6 source files that are out of date
updating environment: 0 added, 6 changed, 0 removed
reading sources... [100%] .generated/psij.launchers.scripts                                                                                                                                 
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] programming                                                                                                                                                        
/mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/batch/cobalt.py:docstring of psij.executors.batch.cobalt.CobaltJobExecutor.generate_submit_script:: WARNING: py:class reference target not found: TextIO
/mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/batch/lsf.py:docstring of psij.executors.batch.lsf.LsfJobExecutor.generate_submit_script:: WARNING: py:class reference target not found: TextIO
/mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/batch/slurm.py:docstring of psij.executors.batch.slurm.SlurmJobExecutor.generate_submit_script:: WARNING: py:class reference target not found: TextIO
/mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/batch/batch_scheduler_executor.py:docstring of psij.executors.batch.batch_scheduler_executor.BatchSchedulerExecutor.generate_submit_script:: WARNING: py:class reference target not found: TextIO
/mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/batch/cobalt.py:docstring of psij.executors.batch.cobalt.CobaltJobExecutor.generate_submit_script:: WARNING: py:class reference target not found: TextIO
/mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/batch/lsf.py:docstring of psij.executors.batch.lsf.LsfJobExecutor.generate_submit_script:: WARNING: py:class reference target not found: TextIO
/mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/batch/pbspro.py:docstring of psij.executors.batch.pbspro.PBSProJobExecutor.generate_submit_script:: WARNING: py:class reference target not found: TextIO
/mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/batch/script_generator.py:docstring of psij.executors.batch.script_generator.SubmitScriptGenerator.generate_submit_script:: WARNING: py:class reference target not found: TextIO
/mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/batch/script_generator.py:docstring of psij.executors.batch.script_generator.TemplatedScriptGenerator.generate_submit_script:: WARNING: py:class reference target not found: TextIO
/mnt/home/merzky/projects/exaworks/psi-j-python/src/psij/executors/batch/slurm.py:docstring of psij.executors.batch.slurm.SlurmJobExecutor.generate_submit_script:: WARNING: py:class reference target not found: TextIO
generating indices... genindex py-modindex done
highlighting module code... [100%] psij.utils                                                                                                                                               
writing additional pages... search done
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build finished with problems, 10 warnings.
Makefile:41: recipe for target 'docs' failed
make: *** [docs] Error 1
```

Do you have a suggestion on what I might be missing?

**Comment by hategan** [comment]
> Do you have a suggestion on what I might be missing?

The `TextIO` errors are something I have seen before with `distutils.version`. I could not find a solution except to blacklist that link since it was not explicitly part of the docstrings, but inferred from the type hints.

I am not sure what causes it and the internets are pretty much crickets on it. What happens if you upgrade sphinx to 4.4.0?

**Comment by andre-merzky** [comment]
> I am not sure what causes it and the internets are pretty much crickets on it. What happens if you upgrade sphinx to 4.4.0?

Alas, sphinx update does not resolve the issue.  Could also not find anything useful yet on the interwebs, will continue to look...

**Comment by hategan** [comment]
I added doc builds to the checks to see if maybe this is python version related.

File: src/psij/launchers/__init__.py [hunk 1]
```diff
@@ -2,10 +2,9 @@
 
 Launcher implementations must be in this package in order for them to be registered automatically.
 Alternatively, you may register a launcher implementation using
-:func:`psij.launchers.launcher.Launcher.register_launcher`.
+:func:`psij.launcher.Launcher.register_launcher`.
```
Comment (jameshcorbett) [inline src/psij/launchers/__init__.py:5]: There are a few changes like this. Shouldn't they go in #159?

**Comment by hategan** [comment]
A bit more on the missing reference errors. They come from intershpinhx attempting to find the relevant classes in the standard python documentation by looking at `https://docs.python.org/3/library/objects.inv`. Attempts to find ,e.g., distutils.version.StrictVersion fail since it's not a documented class and is not in objects.inv (if you really want to take a look at that file, remove the header -- the lines starting with '#' -- in an editor and run `pigz -d -z <objects.inv_no_header>`).

It's unclear why this fails for TextIO, since objects.inv does contain typing.TextIO.

**Comment by hategan** [comment]
Can anybody else test this? I can't reproduce what Andre is seeing.
If this is limited to Andre's machine, I would rather find a workaround for him, such as removing the "-n" argument to sphinx-build in the Makefile.

**Review by andre-merzky — APPROVED** [review]
I could not reproduce my problems on another host, thus approving.  Thanks for your patience!

**Pull request merged**